### ISSUE TITLE
Split up user API into standardized pieces 

### ISSUE THREAD

**Issue by staticfloat** [PumasAI/SimpleChains.jl#60]
The ML model training lifecycle has standardized around three pieces:

1) Forward pass.  This is generally invoked via `model(x)`, although in this repository I think you're more likely to use `model(x, p)`.  Either way is fine for me.

2) Backward pass.  This is generally done through some kind of wrapper function, e.g. `Flux.gradient(() -> loss(model(x), y), p) end`, although a more explicit `gradients(model, p, x, Δ)` would work too.  Note that this piece is what is most bundled up in SimpleChains, as SimpleChains currently groups together the model architecture, weights, loss function, and verification dataset.

3) Optimization/weight update.  This is generally done through some kind of `update!(opt, weights, grads)` call.  This gives the user the ability to modify gradients before passing them to the optimizer, to modify the optimizer itself, etc...

Given these three "verbs", I think most users would be able to easily move their custom training loops to use this, and more importantly, each piece would more or less be independent, which allows for easily incorporating it piece-meal into user's models.  Even if I can only use SimpleChains for 1/3 of my model, cutting down allocations by 1/3 is still significant!

**Comment by rejuvyesh** [comment]
Similar distinction is made in https://github.com/NVlabs/tiny-cuda-nn for example. One can still just use their network with plain pytorch for flexibility rather than their custom loss and optimizer definitions. Should not be too difficult to hook into `ChainRules` to support that?
